### PR TITLE
Introduce PluginInstaller

### PR DESCRIPTION
**Pull request by skrzypo987**
## Release notes

(x) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/server/PluginManager.java [hunk 1]
```diff
@@ -91,6 +91,7 @@
     private final BlockEncodingManager blockEncodingManager;
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/server/PluginManager.java:91]: Improve  the commit message. Maybe something like below?

```
Allow to inject plugin installer

That way when adding new Plugin type we can provide associated with 
plugin installer that we handle it installation without changes in PluginManager.
```

Can you please post a PR where you try to use that new concept for existing plugin types like password manager. The best would be to handle them all in one step.

**Comment by martint** [comment]
What would a such a plugin do if Trino is not aware of it and doesn’t know how to use it? 

Please describe the motivation for this change.

**Comment by kokosing** [comment]
> What would a such a plugin do if Trino is not aware of it and doesn’t know how to use it?

@martint One can use `io.trino.server.Server#getAdditionalModules` to inject support for custom functionality.

Also please notice https://github.com/trinodb/trino/pull/16066 that also presents the value of `PluginInstaller` to Trino itself. Thanks to this `PluginManager` follows open-close principle whenever a new type of the plugin is added.

**Comment by kokosing** [comment]
> But a plugin needs to be inserted somewhere in the flow of a query. 

Yes, for example `SystemSecurityMetadata` can be customized as we are using optional binder. Then custom plugin may affect it is behavior. 

However, my motivation was to allow plugin to affect behavior of other processes than query execution that were added to customized Trino. IMO allowing changes of query behavior should be only allowed only with dedicated SPI. Query execution is too complex to be reasonably customized.

File: core/trino-main/src/main/java/io/trino/server/PluginManager.java [hunk 2]
```diff
@@ -275,4 +282,9 @@ public static PluginClassLoader createClassLoader(String pluginName, List<URL> u
             PluginClassLoader create(String pluginName, List<URL> urls);
         }
     }
+
+    public interface PluginInstaller
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/server/PluginManager.java:286]: extract this class to separate file. It is not implementation detail of PluginManager, it is more an API.

**Review by kokosing — APPROVED** [review]

**Comment by martint** [comment]
> However, my motivation was to allow plugin to affect behavior of other processes than query execution that were added to customized Trino. 

You don’t need a plugin for that. You can just bind additional Guice modules to extend the Trino server if you need additional functionality.

It’s still not clear to me what problem we’re trying to solve.

**Comment by kokosing** [comment]
> You don’t need a plugin for that. You can just bind additional Guice modules to extend the Trino server if you need additional functionality.

Yes. But if you want to have this functionality in the plugin, behind the plugin classloader isolation and as part of the bigger plugin then additional plugins are not enough.

**Pull request merged**